File: golang/array_strings/question_1_1.go
/*
Is Unique:
	Implement an algorithm to determine if a string has all unique characters. What if you
cannot use additional data structures?
*/
package main

import (
	"fmt"
	"os"
	"strings"
)

func main() {
	if len(os.Args) < 0 {
		fmt.Println("You need to inform a string")
		os.Exit(1)
	}

	userString := strings.ToLower(os.Args[1])

	fmt.Printf("Brute-force Solution (%v) -> %v \n", userString, bruteForceSolution(userString))
	fmt.Printf("HashMap Solution (%v) -> %v \n", userString, hashMapSolution(userString))
	fmt.Printf("BitArray Solution (%v) -> %v \n", userString, bitArraySolution(userString))

	os.Exit(0)
}

func bruteForceSolution(userString string) bool {
	/* This is a O(N2) solution without extra space */
	for index := 0; index < len(userString); index++ {
		for secondIndex := index + 1; secondIndex < len(userString); secondIndex++ {
			if userString[index] == userString[secondIndex] {
				return false
			}
		}
	}

	return true
}

func hashMapSolution(userString string) bool {
	/* This is a O(N) solution using an extra space */
	counter := make(map[rune]bool)

	for _, character := range userString {
		if _, found := counter[character]; found {
			return false
		}

		counter[character] = true
	}

	return true
}

func bitArraySolution(userString string) bool {
	checker := 0
	startCodePoint := 'a'

	for i := 0; i < len(userString); i++ {
		mask := 1 << (uint(userString[i]) - uint(startCodePoint))

		sameBits := checker & mask

		if sameBits > 0 {
			return false
		}

		checker = checker | mask
	}

	return true
}

File: c/array_strings/question_1_1.c
/*
    Is Unique:
         Implement an algorithm to determine if a string has all unique characters. What if you
cannot use additional data structures?
*/

#include <stdio.h>
#include <string.h>
#include <stdbool.h>


bool isUnique(char* string);
bool isUniqueBitArray(char* string);

int main(int argc, char const *argv[]) {
    char string[] = "lucass";

    printf("isUnique(%s) -> %u \n", string, isUnique(string));
    printf("isUniqueBitArray(%s) -> %u \n", string, isUniqueBitArray(string));
    return 0;
}

bool isUnique(char *string) {
    int wordLength = (int) strlen(string);

    for (int i = 0; i < wordLength; i++) {
        for (int j = i + 1; j < wordLength; j++) {
            if (string[i] == string[j]) {
                return false;
            }
        }
    }

    return true;
}

bool isUniqueBitArray(char *string) {
    int wordLength = (int) strlen(string);

    int checker = 0;

    for (int i = 0; i < wordLength; i++) {
        int mask = 1 << (string[i] - 'a');

        if (checker & mask > 0) {
            return false;
        }

        checker |= mask;
    }

    return true;
}


